lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js

```diff
@@ -0,0 +1,96 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2018 The Stdlib Authors.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
+var ln = require( '@stdlib/math/base/special/ln' );
+
+
+// MAIN //
+
+/**
+* Computes the inverse hyperbolic cotangent of a number.
+*
+* ## Method
+*
+* Based on
+*
+* ```tex
+* \operatorname{acosh}(x) = \log \left[ x + \sqrt{ x^2 - 1 } \right]
+* ```
+*
+* we have
+*
+* ```tex
+* \operatorname{acosh}(x) = \begin{cases}
+* \log(x) + \tfrac{\ln}{2} & \text{ if x is large } \\
+* \log \left( 2x-\tfrac{1}{\sqrt{x^2-1}+x} \right) & \text{ if } x > 2 \\
+* \operatorname{log1p}\left( x - 1 + \sqrt{ 2 \cdot (x-1) + (x-1)^2 } \right) & \text{ otherwise }
+* \end{cases}
+* ```
+*
+* ## Special Cases
+*
+* ```tex
+* \begin{align*}
+* \operatorname{acosh}(x) &= \mathrm{NaN}\ \text{ if } x < 1 \\
+* \end{align*}
+* ```
+*
+* @param {number} x - input value
+* @returns {(number|NaN)} inverse hyperbolic cotangent (in radians)
+*
+* @example
+* var v = acoth( 1.0 );
+* // returns NaN
+*
+* @example
+* var v = acoth( 2.0 );
+* // returns ~0.5493
+*
+* @example
+* var v = acoth( NaN );
+* // returns NaN
+*/
+function acoth( x ) {
+	if ( isnan( x ) ) {
+		return NaN;
+	}
+
+	if ( x === 1.0 || x === -1.0) {
+		return 0;
+	}
+
+	if ( x >= -1.0 && x <= 1.0 ) {
+		return NaN;
+	}
+
+	if (!isFinite(x)) {
+		return 0;
+	}
+
+	return (ln((x + 1) / x) + ln(x / (x - 1))) / 2;
```

Comment:
This implementation does not seem right to me. Is this still a work-in-progress?

Reply:
I'm no mathematician so I hope I'm not utterly unfit the job! I derived this implementation from [wolframalpha.com](https://www.wolframalpha.com/input/?i=acoth(2)). One other way of writing this I found later in [wikipedia](https://en.wikipedia.org/wiki/Inverse_hyperbolic_functions#Inverse_hyperbolic_cotangent) was `(1/2) * Math.log((x + 1) / (x - 1))`. I confess I didn't benchmark the difference.
For the case `x === 1.0 || x === -1.0` the function would return Infinity, so I thought 0 would be more consistent with other functions.

Reply:
Ah. Okay. Thanks for the clarification.

As an aside, you don't have to be a mathematician to implement mathematical functions; however, I will say that it does take a while to get a "feel" for implementing numerical algorithms. But this is something which is not conferred by credentials, but more by practice.

As a first part of this journey, one thing we rarely do is implement textbook formulas. The reality is that numerical algorithms are often approximations which frequently optimize for particular value regimes. A perhaps more relevant example than `acosh` is [`atanh`](https://github.com/stdlib-js/stdlib/blob/07a705b6ce5aa3147951f737cdb8e4ee5817a991/lib/node_modules/%40stdlib/math/base/special/atanh/lib/atanh.js#L102). Notice how the Wikipedia formulas for both `atanh` and `acoth` are rather similar, but, for `atanh` (derived from FreeBSD), we are not implementing the textbook formula. Instead, we leverage `log1p` for better handling of small values and also circumvent computation entirely for tiny values.

In short, going back to the RFC, in this case, similar to, e.g., Julia, I would recommend that the implementation literally be

```javascript
var atanh = require( '@stdlib/math/base/special/atanh' );

function acoth( x ) {
    return atanh( 1.0/x );
}
```

This will hopefully resolve all the accuracy test failures in this package's unit tests and will obviate the need for any special value consideration. For example, since `atanh( 1.0 ) = +infinity`, the return value of `acoth( 1.0 )` will naturally be zero.